hedera-node/src/main/java/com/hedera/services/contracts/execution/traceability/HederaTracer.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.contracts.execution.traceability;
+
+import static com.hedera.services.contracts.execution.traceability.ContractActionType.CALL;
+import static com.hedera.services.contracts.execution.traceability.ContractActionType.CREATE;
+import static com.hedera.services.contracts.operation.HederaExceptionalHaltReason.INVALID_SOLIDITY_ADDRESS;
+
+import com.hedera.services.state.submerkle.EntityId;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.hyperledger.besu.evm.Code;
+import org.hyperledger.besu.evm.frame.ExceptionalHaltReason;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.frame.MessageFrame.State;
+
+@Singleton
+public class HederaTracer implements HederaOperationTracer {
+
+    private final List<SolidityAction> allActions;
+    private final Deque<SolidityAction> currentActionsStack;
+    private boolean areActionSidecarsEnabled;
+
+    @Inject
+    public HederaTracer() {
+        this.currentActionsStack = new ArrayDeque<>();
+        this.allActions = new ArrayList<>();
+    }
+
+    @Override
+    public void reset(final boolean areActionSidecarsEnabled) {
+        this.currentActionsStack.clear();
+        this.allActions.clear();
+        this.areActionSidecarsEnabled = areActionSidecarsEnabled;
+    }
+
+    @Override
+    public void traceExecution(MessageFrame frame, ExecuteOperation executeOperation) {
+        if (areActionSidecarsEnabled && currentActionsStack.isEmpty()) {
+            trackActionFor(frame, 0);
```

Comment:
Rather than a zero call depth trace I think we should set those up when we construct the tracer (not using dagger).

Or we should have an init call that would populate it before we go into the execution loop.

Reply:
Got it, took the 2nd approach of having an init method performing initialization logic before the EVM execution.